FlattenSequence/distance(from:to:) benchmarks.

oscbyspro opened the pull request:
I wrote some benchmarks for (#71648). I'm new to `/benchmark`, but I read the README and ran the script.

oscbyspro committed:
FlattenSequence/distance(from:to:) benchmarks (#71648).

glessard commented:
@swift-ci please benchmark

glessard commented:
@swift-ci please smoke test

glessard commented on benchmark/single-source/FlattenDistanceFromTo.swift:
```diff
@@ -0,0 +1,58 @@
+  _ iterations: Int
+) {
+  var value = 0 as Int
+  let minor = repeatElement(00000, count: inner)
+  let major = repeatElement(minor, count: outer)
+  let flattened: FlattenSequence = major.joined()
```
These could be at the top level, so that creating them isn't part of the measurement.

oscbyspro replied:
Are you sure you want to move these? It would be a lot of ceremony for `3` integers. Note that `distance(from:to:)` is called `2 * outer^2 * inner^2 * iterations` times. I'm asking just in case you thought they were arrays.

```
▿ Swift.FlattenSequence<Swift.Repeated<Swift.Repeated<Swift.Int>>>
 ▿ _base: Swift.Repeated<Swift.Repeated<Swift.Int>>
  - count: 32
  ▿ repeatedValue: Swift.Repeated<Swift.Int>
   - count: 32
   - repeatedValue: 0
```

oscbyspro replied:
Nevermind. I found a solution I don't hate, then added a few sequential benchmarks for good measure. It's boilerplate-y but symmetrical. You could make it generic, but perhaps manual specialization is preferred in benchmarks?

glessard commented on benchmark/utils/main.swift:
```diff
@@ -89,6 +89,7 @@ import Exclusivity
 import ExistentialPerformance
 import Fibonacci
 import FindStringNaive
+import FlattenDistanceFromTo
```
Please also add `register(FlattenDistanceFromTo.benchmarks)` at line 286 below.
With the PR as-is the benchmark runner still does not know about this addition.

oscbyspro replied:
Sure. The script could use some love too:

> The script will automatically:
>
> 3. Edit main.swift by importing and registering* your new Swift module.
>
> *The ~~cake~~ *auto-registration* is a lie.

oscbyspro committed:
FlattenSequence/distance(from:to:) benchmarks (v2).

1. Registered the new benchmarks in benchmark/utils/main.swift.

oscbyspro committed:
FlattenSequence/distance(from:to:) benchmarks (v3).

1. Parameterized benchmark dependencies, per code review.

oscbyspro committed:
FlattenSequence/distance(from:to:) benchmarks (v4).

1. Added some sequential tests with String as the inner collection.

oscbyspro commented:
I've updated the PR and added a few string-based tests. Here's some results using plain XCTest:

#### Swift 5.9

```
Test Case '-[FlattenTests.FlattenDistanceFromTo testRandomAccess16x16]' passed (0.027 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRandomAccess16x32]' passed (0.116 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRandomAccess32x16]' passed (0.125 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRandomAccess32x32]' passed (0.908 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRepeatString08x08]' passed (0.007 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRepeatString08x16]' passed (0.038 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRepeatString16x08]' passed (0.021 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRepeatString16x16]' passed (0.299 seconds).
```

#### After (#71648)

```
Test Case '-[FlattenTests.FlattenDistanceFromTo testRandomAccess16x16]' passed (0.002 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRandomAccess16x32]' passed (0.003 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRandomAccess32x16]' passed (0.004 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRandomAccess32x32]' passed (0.016 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRepeatString08x08]' passed (0.002 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRepeatString08x16]' passed (0.015 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRepeatString16x08]' passed (0.013 seconds).
Test Case '-[FlattenTests.FlattenDistanceFromTo testRepeatString16x16]' passed (0.108 seconds).
```

oscbyspro committed:
FlattenSequence/distance(from:to:) benchmarks (v5).

1. Replaced occurrences of Repeated<T> with Array<T>.

oscbyspro commented:
I suppose you may write an `O(1)` algorithm if all inner collections have the same number of elements, so I replaced occurrences of `Repeated<T>` with `Array<T>`. Array may also be a more interesting test subject, as it’s more ubiquitous :man_shrugging:

glessard commented:
@swift-ci please benchmark

glessard commented:
@swift-ci please benchmark

oscbyspro committed:
FlattenSequence/distance(from:to:) benchmarks (v6).

1. The auto-generated [BenchmarkInfo] was given the wrong name.

oscbyspro commented:
I looked at the logs and found that the auto-generated [BenchmarkInfo] was given the wrong name:

```
/Users/ec2-user/jenkins/workspace/swift-PR-macos-perf/swift/benchmark/utils/main.swift:286:32: error: value of type '[BenchmarkInfo]' has no member 'benchmarks'
register(FlattenDistanceFromTo.benchmarks)
         ~~~~~~~~~~~~~~~~~~~~~ ^~~~~~~~~~
```

glessard commented:
@swift-ci please benchmark

glessard commented:
| ✅    | Benchmark Check Report                                       |
| ---- | ------------------------------------------------------------ |
| ⛔️⏱   | `FlattenDistanceFromTo.Array.String.08.16` execution took at least 46308 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.String.08.16` by a factor of 64 (100), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.String.08.16` |
| ⛔️⏱   | `FlattenDistanceFromTo.Array.Array.32.16` execution took at least 514988 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.Array.32.16` by a factor of 1024 (1000), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.Array.32.16` |
| ⚠️⏱   | `FlattenDistanceFromTo.Array.String.08.08` execution took at least 4786 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.String.08.08` by a factor of 8 (10), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.String.08.08` |
| ⛔️⏱   | `FlattenDistanceFromTo.Array.String.16.08` execution took at least 37444 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.String.16.08` by a factor of 64 (100), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.String.16.08` |
| ⛔️⏱   | `FlattenDistanceFromTo.Array.Array.16.16` execution took at least 64735 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.Array.16.16` by a factor of 128 (100), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.Array.16.16` |
| ⛔️⏱   | `FlattenDistanceFromTo.Array.Array.16.32` execution took at least 306107 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.Array.16.32` by a factor of 512 (1000), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.Array.16.32` |
| ⛔️⏱   | `FlattenDistanceFromTo.Array.String.16.16` execution took at least 368676 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.String.16.16` by a factor of 512 (1000), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.String.16.16` |
| ⛔️⏱   | `FlattenDistanceFromTo.Array.Array.32.32` execution took at least 2418596 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.Array.32.32` by a factor of 4096 (10000), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.Array.32.32` |

glessard commented:
It looks like they'd still be a long-running after the performance improvement.

oscbyspro committed:
FlattenSequence/distance(from:to:) benchmarks (v7).

The benchmarks took too long to perform, so I reduced the size of each payload.

oscbyspro commented:
I cut the payload dimensions into halves and quarters.

glessard commented:
@swift-ci please benchmark

glessard commented:
✅  | Benchmark Check Report
---|--- <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.Array.08.04` <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.Array.04.04` <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.String.04.08`
⛔️⏱ | `FlattenDistanceFromTo.Array.String.04.04` has setup overhead of 9 μs (7.6%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.String.04.04`
⚠️⏱ | `FlattenDistanceFromTo.Array.Array.08.08` execution took at least 2021 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.Array.08.08` by a factor of 4 (10), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.Array.08.08` <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.Array.04x08`
⚠️⏱ | `FlattenDistanceFromTo.Array.String.08.08` execution took at least 5171 μs. <br><sub> Decrease the workload of `FlattenDistanceFromTo.Array.String.08.08` by a factor of 8 (10), to be less than 1000 μs. <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.String.08.08` <br><sub> unable to compute memory footprint of `FlattenDistanceFromTo.Array.String.08.04`

glessard commented:
Two mysterious issues remaining in the benchmarks:
- it complains about the setup time for `FlattenDistanceFromTo.Array.String.04.04`, though it's no different than the others.
- it complains about not figuring out the memory footprint in most cases.
These seem like non-issues to me.

glessard commented:
@swift-ci please smoke test
